### ISSUE TITLE
Grafana Dashboards have 'dips'

### ISSUE THREAD

**Issue by AlexLamond** [unpoller/unpoller#755]
![image](https://github.com/user-attachments/assets/9a273114-7199-4e33-8fab-80dc65259ea5)

After setting up Unpoller via Docker, with a Unifi Express I am finding that the charts on Grafana are shown with 'dips'. I would expect these lines to be nice and smooth, but sadly they keep reporting zero

**Comment by davidnewhall** [comment]
Increase the poll interval.

**Comment by AlexLamond** [comment]
Maybe I'm just blind, but what is the Docker ENV tag to do that?

**Comment by drbrain** [comment]
There appears to be [`UP_INFLUXDB_INTERVAL`](https://unpoller.com/docs/install/configuration#influxdb) for InfluxDB, but [nothing for Prometheus](https://unpoller.com/docs/install/configuration#prometheus)

I'm using prometheus and have this same problem.

**Comment by davidnewhall** [comment]
Prometheus scrape interval?

**Comment by drbrain** [comment]
Increasing the prometheus scrape interval (how often prometheus requests metrics from unpoller) does not appear to make the unpoller prometheus exporter fetch metrics from unifi more often.

I haven't dug into how `UP_INFLUXDB_INTERVAL` is implemented, but maybe increasing it will also increase the update frequency for the prometheus exporter?

**Comment by davidnewhall** [comment]
>  unpoller appears to fetch metrics on its own internal interval and export the cached copy.

That's certainly not how I wrote the code and I doubt @platinummonkey changed that logic. It fetches metrics during the scrape request.

**Comment by drbrain** [comment]
I rebuilt `unpoller` from 96a9259eb9e9ba138a39d6cf2691d4240c8b57e7 (today's `master`).

My prometheus scrape interval is 15s

The top panel in this screenshot is from `telegraf` scraping SNMP from my unifi switch directly.

The bottom panel in this screenshot is `unpoller`. It scrapes a UCK G2 Plus running Unifi OS v4.0.18 and Network 8.4.62.

I suspect both to look the same, but `unpoller` only has new metrics about once per minute, sometimes less.

<img width="1324" alt="Prometheus Time Series Collection and Processing Server 2024-10-08 12-07-42" src="https://github.com/user-attachments/assets/4f0bf300-63af-426f-9b69-b0cead0e07df">

**Comment by davidnewhall** [comment]
You're reporting a completely different behavior than this original issue. This is normal. The API doesn't update as often as the SNMP values.

**Comment by drbrain** [comment]
My understanding of prometheus, sampling, and grafana dashboard constructing lead me to believe the two behaviors are tied to one another. Zero values exist when `rate()` is used with a range vector that is shorter than the update interval, but there is no way to configure `unpoller` to fetch data more quickly.

Upthread you wrote:

> Increase the poll interval.

Then:

> It fetches metrics during the scrape request.

Both of which are at odds with:

> The API doesn't update as often as the SNMP values.

So it appears that `unpoller` cannot sample often enough for prometheus `rate()` and friends to avoid returning 0 values with a range vector of less than about `[3m]`. As `rate()` and range vectors are the tools used to build dashboards that don't drop to 0.

I understand `unpoller` may not have control of how often the API updates.

Therefore I expect the configuration documentation scrape interval or update interval shorter than one minute is not going to result in improved resolution.

I also expect that the pre-built dashboards would have a minimum interval set to match.

I also expect that the slow API update rate is why the pre-computed metric `unpoller_device_port_receive_rate_bytes` also shows a 0 byte receive rate.

I understand we may have no way to win here, but I haven't seen any documentation about this limitation, and I believe that drove the initial reporter to create this issue.

<img width="1561" alt="Prometheus Time Series Collection and Processing Server 2024-10-08 12-35-15" src="https://github.com/user-attachments/assets/d2ced1d3-bafb-496b-9ea9-265ced5e8422">

**Comment by drbrain** [comment]
Does documentation need to be written to address the issue with scrape intervals, or is there another way to resolve this issue?

**Comment by platinummonkey** [comment]
> Does documentation need to be written to address the issue with scrape intervals, or is there another way to resolve this issue?

Yes this would be the way to do this. Docs website is available here https://github.com/unpoller/unpoller.github.io/tree/master/documentation
